src/dune_tui/dune_tui.ml

```diff
@@ -2,36 +2,56 @@ open Import
 open Lwd.O
 module Unicode = Drawing.Unicode
 
-let create () = Term.create ~nosig:false ~output:Unix.stderr ()
 let bytes = Bytes.make 64 '0'
 let sigcont_pipe = lazy (Unix.pipe ~cloexec:true ())
 
-let term =
-  let setup =
-    lazy
-      (Unix.set_nonblock (Lazy.force sigcont_pipe |> fst);
-       let term = ref (create ()) in
-       Sys.set_signal Sys.sigcont
-       @@ Sys.Signal_handle
-            (fun _ ->
-              Term.release !term;
-              term := create ();
-              assert (1 = Unix.single_write (Lazy.force sigcont_pipe |> snd) bytes 0 1));
-       let rec old =
-         lazy
-           (Sys.signal Sys.sigtstp
-            @@ Sys.Signal_handle
-                 (fun i ->
-                   Term.release !term;
-                   match Lazy.force old with
-                   | Sys.Signal_handle f -> f i
-                   | _ -> Unix.kill (Unix.getpid ()) Sys.sigstop))
-       in
-       ignore (Lazy.force old);
-       term)
-  in
-  fun () -> !(Lazy.force setup)
-;;
+module Terminal_manager = struct
+  let term_ref : Notty_unix.Term.t option ref = ref None
+  let old_sigtstp_handler = ref Sys.Signal_default
+
+  let get_term () =
+    match !term_ref with
+    | Some term -> term
+    | None -> Code_error.raise "Terminal is not enabled." []
+  ;;
+
+  let disable () =
+    match !term_ref with
+    | None -> ()
+    | Some term ->
+      Term.release term;
+      term_ref := None;
+      Sys.set_signal Sys.sigtstp !old_sigtstp_handler;
+      Sys.set_signal Sys.sigcont Sys.Signal_default
+  ;;
+
+  let rec enable () =
+    match !term_ref with
+    | Some _ -> ()
+    | None ->
+      let term = Term.create ~nosig:false ~output:Unix.stderr () in
+      term_ref := Some term;
+      (* Install signal handlers for user-initiated suspend/resume.
+         The SIGCONT handler should re-enable the terminal. *)
+      let sigcont_pipe_r, sigcont_pipe_w = Lazy.force sigcont_pipe in
+      Unix.set_nonblock sigcont_pipe_r;
+      old_sigtstp_handler
+      := Sys.signal Sys.sigtstp
+         @@ Sys.Signal_handle
+              (fun i ->
+                disable ();
+                match !old_sigtstp_handler with
+                | Sys.Signal_handle f -> f i
+                | _ -> Unix.kill (Unix.getpid ()) Sys.sigstop);
+      Sys.set_signal Sys.sigcont
+      @@ Sys.Signal_handle
+           (fun _ ->
+             if not (Option.is_some !term_ref)
+             then (
+               let _ = enable () in
```

Comment:
Why do we set a signal handler from inside a handler? This code could use some comments.